### PR TITLE
Integrated Auth0 service

### PR DESCRIPTION
**Pull request by YazidZidane**
- Integrated Auth0, now the user can login or signup at Auth0 page.
- Add a popup for login or signup, login/logout button on the header.
- Added user profile page.
- Added user data to redux store.
- Integrated with user history api to save the field user select in Location page. Now the logged user could save, delete, update, change name of multiple fields, fields can be point or polygon.

### CODE REVIEW

**Comment by mikahpinegar** [comment]
can you only show one modal at a time? i don't like the overlapping modals
snip1

**Comment by mikahpinegar** [comment]
also, if i change my location the zone dropdown isn't updating properly anymore. ex choose NC and then move the marker down towards Wilmington. it should change to zone 8a but is showing an empty dropdown

**Comment by mikahpinegar** [comment]
could you also add a text box for field name and then save that via the label in the api?

File: src/pages/Header/Header.js [hunk 1]
```diff
@@ -48,6 +73,14 @@ const Header = () => {
         <NavLink to="/feedback" activeClassName="active">
           FEEDBACK
         </NavLink>
+        <span className="line" />
+        {isAuthenticated ? (
+          <LogoutButton />
+        ) : (
+          <a href="/" onClick={handleLogin}>
```
Comment (Adamws33) [inline src/pages/Header/Header.js:80]: Shouldnt this be the login button component? if its just a handler button it should be a <Button> tag.

File: src/components/Auth/Buttons/LogoutButton.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    // dispatch(userLogout());
+    logout({
+      logoutParams: {
+        returnTo: window.location.origin,
+      },
+    }).then(dispatch(userLogout()));
+  };
+
+  return (
+    <a href="/" onClick={handleLogout}>
```
Comment (Adamws33) [inline src/components/Auth/Buttons/LogoutButton.js:20]: Should use a <Button> tag unless there is a specific reason its not.  see MyCoverCropReset.js

File: src/components/Auth/Buttons/SignUpButton.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  const handleSignUp = async () => {
+    await loginWithRedirect({
+      authorizationParams: {
+        screen_hint: 'signup',
+      },
+    });
+  };
+
+  return (
```
Comment (Adamws33) [inline src/components/Auth/Buttons/SignUpButton.js:16]: The three buttons here Login, Logout, and Signup should probably be combined in some way maybe login and signup in one button and logout separate or just a single button that is reused and uses props to determine its text and functionality.

File: src/components/Auth/AuthButton/AuthButton.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        return handleSignUp();
+      case 'Logout':
+        return handleLogout();
+      default:
+        return onClickCallback();
+    }
+  };
+
+  return (
+    <Button variant={variant} onClick={handleClick} component={isAnchor ? 'a' : null}>
```
Comment (Adamws33) [inline src/components/Auth/AuthButton/AuthButton.js:47]: we need to remove the "isAnchor" prop and never set the component to an anchor tag.  we can replace that with a "from" prop that can be used in some inline styling to remove borders and set the font to black.  i also think we need some condiditonal styling to change the color of the logout to a red button and have the button be green when its login or sign up.  @mikahpinegar thoughts?

File: src/components/Auth/Auth0ProviderWithHistory/Auth0ProviderWithHistory.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import { Auth0Provider } from '@auth0/auth0-react';
+import React from 'react';
+import { useHistory } from 'react-router-dom';
+
+const Auth0ProviderWithHistory = ({ children }) => {
+  const history = useHistory();
+
+  const domain = process.env.REACT_APP_AUTH0_DOMAIN;
```
Comment (mikahpinegar) [inline src/components/Auth/Auth0ProviderWithHistory/Auth0ProviderWithHistory.js:8]: move these into keys.js like the other env vars

File: src/components/Auth/AuthButton/AuthButton.js (hunk 4 above)
Comment (mikahpinegar) [reply src/components/Auth/AuthButton/AuthButton.js:47]: yeah i agree with all that. the nav links are needed because they route to a different page on click. how about adding a prop to the button component called isNavButton which will decide whether to return NavLink or Button? then use that component for everything

File: src/pages/Location/Location.js [hunk 6]
```diff
@@ -57,8 +81,54 @@ const LocationComponent = () => {
...
+  const [drawFields, setDrawFields] = useState([]);
+
+  const userFieldsRef = useRef(userFieldRedux ? [...userFieldRedux.data] : []);
+
+  const { isAuthenticated } = useAuth0();
+
+  // console.log('userFieldRedux', userFieldRedux);
+  // console.log('selectedUserField', selectedUserField)
+
+  const getArea = () => {
```
Comment (mikahpinegar) [inline src/pages/Location/Location.js:102]: why is this needed? the bit map already returns area right?

File: src/pages/Location/Location.js [hunk 7]
```diff
@@ -260,6 +344,129 @@ const LocationComponent = () => {
...
+        // console.log('update', polygon);
+        const geoCollection = buildGeometryCollection(point.geometry, polygon.geometry, selectedUserField);
+        userFieldsRef.current = userFieldsRef.current.map((f) => {
+          if (f.label === selectedUserField) return geoCollection;
+          return f;
+        });
+      }
+      if (actionType === 'delete') {
+        // add a delete flag so when saving the code knows delete which field
+        userFieldsRef.current = userFieldsRef.current.map((f) => {
```
Comment (Adamws33) [inline src/pages/Location/Location.js:428]: all of these maps should have the "f" prop from the map changed to something more readable.  i suggest "userFieldRef" or "userField" ill leave comments on the others i found too.

File: src/pages/Location/Location.js [hunk 8]
```diff
@@ -260,6 +344,129 @@ const LocationComponent = () => {
...
+        setSelectedUserField(fieldName);
+      }
+      if (actionType === 'update') {
+        // Only supports polygon updates
+        const { longitude, latitude } = selectedToEditSite;
+        const point = buildPoint(longitude, latitude, selectedUserField);
+        const polygon = currArea.features.slice(-1)[0];
+        // console.log('update', polygon);
+        const geoCollection = buildGeometryCollection(point.geometry, polygon.geometry, selectedUserField);
+        userFieldsRef.current = userFieldsRef.current.map((f) => {
```
Comment (Adamws33) [inline src/pages/Location/Location.js:421]: all of these maps should have the "f" prop from the map changed to something more readable. I suggest "userFieldRef" or "userField" ill leave comments on the others i found too.

File: src/pages/Location/Location.js [hunk 9]
```diff
@@ -260,6 +344,129 @@ const LocationComponent = () => {
...
+        // add a delete flag so when saving the code knows delete which field
+        userFieldsRef.current = userFieldsRef.current.map((f) => {
+          if (f.label === selectedUserField) {
+            // if this field have not been saved to the backend, directly delete it from the userFields.
+            if (!f.id) return null;
+            return { ...f, delete: true };
+          }
+          return f;
+        });
+        userFieldsRef.current = userFieldsRef.current.filter((f) => f !== null);
```
Comment (Adamws33) [inline src/pages/Location/Location.js:436]: all of these maps should have the "f" prop from the map changed to something more readable. i suggest "userFieldRef" or "userField" ill leave comments on the others i found too.

File: src/pages/Location/Location.js [hunk 10]
```diff
@@ -260,6 +344,129 @@ const LocationComponent = () => {
...
+        // find next available field in current fields array
+        const nextAvailableField = userFieldsRef.current.find((f) => f.delete !== true);
```
Comment (Adamws33) [inline src/pages/Location/Location.js:438]: all of these maps should have the "f" prop from the map changed to something more readable. i suggest "userFieldRef" or "userField" ill leave comments on the others i found too.

File: src/pages/Location/Location.js [hunk 11]
```diff
@@ -260,6 +344,129 @@ const LocationComponent = () => {
...
+      if (actionType === 'updateName') {
+        const { prevName } = fieldDialogState;
+        const errText = fieldNameValidation(fieldName, userFieldsRef.current);
+        if (errText !== '') {
+          setFieldDialogState({ ...fieldDialogState, error: true, errorText: errText });
+          return;
+        }
+        const newField = {
+          type: 'Feature',
+          geometry: userFieldsRef.current.filter((f) => f.label === prevName)[0].geometry,
```
Comment (Adamws33) [inline src/pages/Location/Location.js:450]: all of these maps should have the "f" prop from the map changed to something more readable. i suggest "userFieldRef" or "userField" ill leave comments on the others i found too.

File: src/pages/Location/UserFieldList/UserFieldList.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+          <Select
+            variant="filled"
+            style={{
+              width: '100%',
+              textAlign: 'left',
+            }}
+            value={field}
+            onChange={handleChange}
+          >
+            {userFields.map((f, i) => {
```
Comment (Adamws33) [inline src/pages/Location/UserFieldList/UserFieldList.js:46]: a map update to enhance readability

File: src/shared/constants.js [hunk 13]
```diff
@@ -893,3 +893,86 @@ export const cropDataFormatter = (cropData = [{}]) => {
     return val;
   });
 };
+
+export const apiServerUrl = process.env.REACT_APP_API_SERVER_URL;
```
Comment (mikahpinegar) [inline src/shared/constants.js:897]: i don't think this needs to be an env

**Review by mikahpinegar — CHANGES_REQUESTED** [review]

File: src/pages/Location/UserFieldDialog/UserFieldDialog.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+  } = fieldDialogState;
+  return (
+    <Dialog sx={{ m: 0, p: 2 }} open={open}>
+      {actionType === 'add' && (
+      <>
+        <DialogTitle>
+          You need to give this field a nickname
+        </DialogTitle>
+        <DialogContent>
+          <TextField
```
Comment (Adamws33) [inline src/pages/Location/UserFieldDialog/UserFieldDialog.js:23]: the 2 TextFields here are the same could they be set to a single const

File: src/pages/Location/UserFieldDialog/UserFieldDialog.js [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+            helperText={errorText}
+            variant="standard"
+            label="Field Nickname"
+            value={fieldName}
+            onChange={(e) => setFieldDialogState({ ...fieldDialogState, fieldName: e.target.value })}
+          />
+        </DialogContent>
+      </>
+      )}
+      {actionType === 'update' && (
```
Comment (Adamws33) [inline src/pages/Location/UserFieldDialog/UserFieldDialog.js:35]: I think we just need to use the actionType to set a const string and then just bring that string into the DialogTitle and create a conditional just to display the TextField or not.  this will reduce a few lines and conditionals.

File: src/pages/Location/Location.js (hunk 6 above)
Comment (YazidZidane) [reply src/pages/Location/Location.js:102]: This function sets the initFeatures for the map, whenever the user selected Fields changed, it sets the selected Field to the map.

Comment (mikahpinegar) [reply src/pages/Location/Location.js:103]: ok then can you rename it to something that makes more sense? like updateGeometry perhaps?

File: src/pages/Profile/Profile.js [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+              <span style={{
+                fontSize: '1.3rem',
+              }}
+              >
+                {user.email}
+              </span>
+            </div>
+          </div>
+        ) : (
+          <div>You have not been logged in !</div>
```
Comment (mikahpinegar) [inline src/pages/Profile/Profile.js:56]: remove space before exclamation

**Review by mikahpinegar — CHANGES_REQUESTED** [review]

File: src/pages/Location/Location.js [hunk 17]
```diff
@@ -57,8 +82,54 @@ const LocationComponent = () => {
   const [selectedToEditSite, setSelectedToEditSite] = useState({});
   const [handleConfirm, setHandleConfirm] = useState(false);
   const [locZipCode, setLocZipCode] = useState();
+  const [currArea, setCurrArea] = useState([]);
```
[the remark this answers is not in the document]
Comment (mikahpinegar) [reply src/pages/Location/Location.js:85]: or currGeometry, setCurrGeometry

**Review by mikahpinegar — APPROVED** [review]
